core-plugins/src/main/java/co/cask/hydrator/plugin/common/StructuredToAvroTransformer.java

```diff
@@ -63,7 +63,12 @@ public GenericRecord transform(StructuredRecord structuredRecord) throws IOExcep
       if (schemaField == null) {
         throw new IllegalArgumentException("Input record does not contain the " + fieldName + " field.");
       }
-      recordBuilder.set(fieldName, convertField(structuredRecord.get(fieldName), schemaField.getSchema()));
+      if (schemaField.getSchema().getType() == co.cask.cdap.api.data.schema.Schema.Type.RECORD) {
```

Comment:
this is not a good way to fix it. convertField already has a switch statement on the field type, you shouldn't duplicate that logic here, and shouldn't be creating a new transformer for each recursive call.

Instead, just remove the outputAvroSchema variable completely. The avro schema should always be calculated from the  cdap schema.